packages/hebao_v1/contracts/stores/GuardianStore.sol

```diff
@@ -0,0 +1,307 @@
+            Data.Guardian memory g = w.guardians[i];
+            if (_isPendingAddition(g)) {
+                w.guardians[i].status = uint8(Data.GuardianStatus.REMOVE);
+                w.guardians[i].timestamp = 0;
+            }
+            if (_isPendingRemoval(g)) {
+                w.guardians[i].status = uint8(Data.GuardianStatus.ADD);
+                w.guardians[i].timestamp = 0;
+            }
+        }
+        _cleanRemovedGuardians(wallet);
+    }
+
+    function addGuardian(
+        address wallet,
+        address addr,
+        uint    validSince,
+        bool    alwaysOverride
+        )
+        public
+        onlyWalletModule(wallet)
+        returns (uint)
+    {
+        require(validSince >= block.timestamp, "INVALID_VALID_SINCE");
+        require(addr != address(0), "ZERO_ADDRESS");
+
+        Wallet storage w = wallets[wallet];
+        uint pos = w.guardianIdx[addr];
+
+        if(pos == 0) {
+            // Add the new guardian
+            Data.Guardian memory g = Data.Guardian(
+                addr,
+                uint8(Data.GuardianStatus.ADD),
+                validSince.toUint64()
+            );
+            w.guardians.push(g);
+            w.guardianIdx[addr] = w.guardians.length;
+
+            _cleanRemovedGuardians(wallet);
+            return validSince;
+        }
+
+        Data.Guardian memory g = w.guardians[pos - 1];
+
+        if (_isRemoved(g)) {
+            w.guardians[pos - 1].status = uint8(Data.GuardianStatus.ADD);
+            w.guardians[pos - 1].timestamp = validSince.toUint64();
+            return validSince;
+        }
+
+        if (_isPendingRemoval(g)) {
+            w.guardians[pos - 1].status = uint8(Data.GuardianStatus.ADD);
+            w.guardians[pos - 1].timestamp = 0;
+            return 0;
+        }
+
+        if (_isPendingAddition(g)) {
+            if (!alwaysOverride) return g.timestamp;
+
+            w.guardians[pos - 1].timestamp = validSince.toUint64();
+            return validSince;
+        }
+
+        require(_isAdded(g), "UNEXPECTED_RESULT");
+        return 0;
+    }
+
+    function removeGuardian(
+        address wallet,
+        address addr,
+        uint    validUntil,
+        bool    alwaysOverride
+        )
+        public
+        onlyWalletModule(wallet)
+        returns (uint)
+    {
+        require(validUntil >= block.timestamp, "INVALID_VALID_UNTIL");
+        require(addr != address(0), "ZERO_ADDRESS");
+
+        Wallet storage w = wallets[wallet];
+        uint pos = w.guardianIdx[addr];
+        require(pos > 0, "GUARDIAN_NOT_EXISTS");
+
+        Data.Guardian memory g = w.guardians[pos - 1];
+
+        if (_isAdded(g)) {
+            w.guardians[pos - 1].status = uint8(Data.GuardianStatus.REMOVE);
+            w.guardians[pos - 1].timestamp = validUntil.toUint64();
+            return validUntil;
+        }
+
+        if (_isPendingAddition(g)) {
+            w.guardians[pos - 1].status = uint8(Data.GuardianStatus.REMOVE);
```

Comment:
remove pending guardian immedately here.

Reply:
Isn't that what happens here? It's just not removed from the list which I think is fine.